state/protocol/protocol_state/kvstore/statemachine.go

```diff
@@ -0,0 +1,104 @@
+package kvstore
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/state/protocol/protocol_state"
+)
+
+// ProcessingStateMachine is a dedicated structure that encapsulates all logic for evolving KV store, based on the content
+// of a new block.
+// ProcessingStateMachine processes a subset of service events that are relevant for the KV store, and ignores all other events.
+// Each relevant event is validated before it is applied to the KV store.
+// All updates are applied to a copy of parent KV store, so parent KV store is not modified.
+// A separate instance should be created for each block to process the updates therein.
+type ProcessingStateMachine struct {
```

Comment:
```suggestion
type StateMachine struct {
```
Would suggest to just call this `kvstore.StateMachine`, as it is the only state machine in this package.